src/js/port_handler.js

```diff
@@ -23,86 +22,135 @@ const PortHandler = new function () {
     EventBus.$on('ports-input:change', this.onChangeSelectedPort.bind(this));
 
-    serial.addEventListener("addedDevice", this.check_serial_devices.bind(this));
-    serial.addEventListener("removedDevice", this.check_serial_devices.bind(this));
+    serial.addEventListener("addedDevice", (event) => this.addedSerialDevice(event.detail));
+    serial.addEventListener("removedDevice", (event) => this.removedSerialDevice(event.detail));
+
+    if (this.usbCheckLoop) {
+        clearTimeout(this.usbCheckLoop);
+    }
+
+    if (!this.portAvailable) {
+        this.check_usb_devices();
+    }
 
-    this.reinitialize();    // just to prevent code redundancy
+    if (!this.dfuAvailable) {
+        this.addedSerialDevice();
+    }
 };
 
 PortHandler.setShowVirtualMode = function (showVirtualMode) {
     this.showVirtualMode = showVirtualMode;
-    this.selectActivePort();
+    this.selectBestActivePort();
 };
 
 PortHandler.setShowManualMode = function (showManualMode) {
     this.showManualMode = showManualMode;
-    this.selectActivePort();
+    this.selectBestActivePort();
 };
 
-PortHandler.reinitialize = function () {
-    this.initialPorts = false;
-
-    this.showAllSerialDevices = getConfig('showAllSerialDevices').showAllSerialDevices;
+PortHandler.addedSerialDevice = function (device) {
+    this.updateCurrentPortsList()
+    .then(() => {
+        const selectedPort = this.selectBestActivePort(device);
+        if (!device || selectedPort === device.path) {
+            // Send this event when the port handler auto selects a new device
+            EventBus.$emit('port-handler:auto-select-device', selectedPort);
+        }
+    });
+};
 
-    this.check();   // start listening, check after TIMEOUT_CHECK ms
+PortHandler.removedSerialDevice = function (device) {
+    this.updateCurrentPortsList()
+    .then(() => {
+        if (this.portPicker.selectedPort === device.path) {
+            this.selectBestActivePort();
+        }
+    });
 };
 
-PortHandler.check = function () {
-    const self = this;
+PortHandler.onChangeSelectedPort = function(port) {
+    this.portPicker.selectedPort = port;
+};
 
-    if (!self.port_available) {
-        self.check_usb_devices();
-    }
+PortHandler.updateCurrentPortsList = function () {
+    return serial.getDevices()
+    .then((ports) => {
+        ports = this.sortPorts(ports);
+        this.currentPorts = ports;
+    });
+};
 
-    if (!self.dfu_available) {
-        self.check_serial_devices();
-    }
+PortHandler.sortPorts = function(ports) {
+    return ports.sort(function(a, b) {
+        return a.path.localeCompare(b.path, window.navigator.language, {
+            numeric: true,
+            sensitivity: 'base',
+        });
+    });
+};
 
+PortHandler.askSerialPermissionPort = function() {
+    serial.requestPermissionDevice();
 };
 
-PortHandler.check_serial_devices = function () {
-    const self = this;
+PortHandler.selectBestActivePort = function(suggestedDevice) {
```

Comment:
What was wrong with `selectActivePort`?

Reply:
Nothing, I prefered this, but I will change it again ;)